File: 과제1c - PYTHON으로 파서 구현/Project1c/LiteralTable.py
class LiteralTable:
    def __init__(self):
        '''
        클래스를 초기화한다.
        '''
        self.literalList = []
        self.locationList = []

    def putLiteral(self,literal,location):
        '''
        새로운 Literal을 table에 추가하는 함수이다. 중복된 literal은 putLiteral로 입력될 수 없다. literal의 주소를 변경하려면 modifyLital()를 이용한다.

        literal : 새로 추가되는 literal의 label\n
        location : 해당 literal의 주소값
        '''
        tmpliteral = literal[1:]
        if not tmpliteral in self.literalList:
            self.literalList.append(tmpliteral)
            self.locationList.append(location)
    
    def modifyLiteral(self, literal, newLocation):
        '''
        기존에 존재하는 literal 값에 대해서 해당 literal의 주소값을 변경한다.

        literal : 주소값의 변경을 원하는 literal의 label\n
        newLocation : 새로 바꾸고자 하는 주소값
        '''
        if literal in self.literalList:
            index = self.literalList.index(literal)
            self.locationList[index] = newLocation

    def search(self, literal):
        '''
        인자로 전달된 literal이 어떤 주소를 지칭하는지 알려준다.

        literal : 검색을 원하는 literal의 label\n
        return : literal이 가지고 있는 주소값. 해당 literal이 없을 경우 -1 리턴
        '''
        if literal in self.literalList:
            index = self.literalList.index(literal)
            return self.locationList[index]
        else:
            return -1

    def getAddress(self, index):
        '''
        주어진 index에 해당하는 address을 리턴한다.
        
        index : address을 검색할 인덱스\n
        return : index에 해당하는 address
        '''
        return self.locationList[index]

    def getLiteral(self, index):
        '''
        주어진 index에 해당하는 literal을 리턴한다.
        
        index : literal을 검색할 인덱스\n
        return : index에 해당하는 literal
        '''
        return self.literalList[index]


    def addrLiteralTab(self, locctr):
        '''
        literal들의 주소를 넣어주는 함수이다.

        locctr : 현재 프로그램의 주소\n
        return : literal들의 주소를 넣어주고 증가한 주소를 리턴
        '''
        for i in range(0,len(self.literalList)):
            if -2 in self.locationList:
                self.modifyLiteral(self.literalList[i],locctr)
                if self.literalList[i][0] == 'C' or self.literalList[i][0] == 'c': # literal이 'C'인지 확인
                    tmpStr = self.literalList[i]
                    tmpStr = tmpStr[2:len(tmpStr)-1]
                    locctr += len(tmpStr) #char개수만큼 locctr 증가
                elif self.literalList[i][0] == 'X' or self.literalList[i][0] == 'x': #literal이 'X'인지 확인
                    tmpStr = self.literalList[i]
                    tmpStr = tmpStr[2:len(tmpStr)-1]
                    locctr += int(len(tmpStr)/2) #char개수/2만큼 locctr 증가
        return locctr

    def getSize(self):
        '''
        literaltable의 크기를 알려주는 함수이다.

        return : literal table의 크기
        '''
        return len(self.literalList)

File: 과제1b - JAVA로 파서 구현/Project1b/src/LiteralTable.java
import java.util.ArrayList;

/**
 * literal과 관련된 데이터와 연산을 소유한다.
 * section 별로 하나씩 인스턴스를 할당한다.
 */
public class LiteralTable {
	ArrayList<String> literalList;
	ArrayList<Integer> locationList;
	// 기타 literal, external 선언 및 처리방법을 구현한다.
	
	public LiteralTable() {
		literalList = new ArrayList<>();
		locationList = new ArrayList<>();
	}
	
	/**
	 * 새로운 Literal을 table에 추가한다.
	 * @param literal : 새로 추가되는 literal의 label
	 * @param location : 해당 literal이 가지는 주소값
	 * 주의 : 만약 중복된 literal이 putLiteral을 통해서 입력된다면 이는 프로그램 코드에 문제가 있음을 나타낸다. 
	 * 매칭되는 주소값의 변경은 modifyLiteral()을 통해서 이루어져야 한다.
	 */
	public void putLiteral(String literal, int location) {
		String tmpInst = new String(literal.substring(1));
		literal = tmpInst;
		if(literalList.contains(literal) == false) {
			literalList.add(literal);
			locationList.add(location);
		}
	}
	
	/**
	 * 기존에 존재하는 literal 값에 대해서 가리키는 주소값을 변경한다.
	 * @param literal : 변경을 원하는 literal의 label
	 * @param newLocation : 새로 바꾸고자 하는 주소값
	 */
	public void modifyLiteral(String literal, int newLocation) {
		int index;
		index = literalList.indexOf(literal);
		locationList.set(index, newLocation);
	}
	
	/**
	 * 인자로 전달된 literal이 어떤 주소를 지칭하는지 알려준다. 
	 * @param literal : 검색을 원하는 literal의 label
	 * @return literal이 가지고 있는 주소값. 해당 literal이 없을 경우 -1 리턴
	 */
	public int search(String literal) {
		int address = 0;
		
		if(literalList.contains(literal)) {
			for (int i = 0; i < literalList.size(); i++) {
				if (literal.equals(literalList.get(i))){
					address = locationList.get(i);
					break;
				}
			}
		}
		else {
			address = -1;
		}
		return address;
	}
	
	/**
	 * literal들의 주소를 넣어주는 함수이다
	 * @param locctr : 현재 프로그램의 주소
	 * @return locctr : literal들의 주소를 넣어주고 증가한 주소 반환
	 */
	public int addrLiteralTab(int locctr) {
		String tmpstr;
		int strsize;
		for (int i = 0; i < literalList.size(); i++) {
			if(search(literalList.get(i)) == -2) {
				modifyLiteral(literalList.get(i),locctr);
				if(literalList.get(i).charAt(0) == 'C' || literalList.get(i).charAt(0) == 'c') { //literal이 'C'인지 확인
					tmpstr = new String(literalList.get(i));
					strsize = tmpstr.length();
					tmpstr = tmpstr.substring(2,strsize-1); 
					locctr += tmpstr.length(); //char개수만큼 locctr 증가
				}
				else if(literalList.get(i).charAt(0) == 'X' || literalList.get(i).charAt(0) == 'x') { //literal이 'X"인지 확인
					tmpstr = new String(literalList.get(i));
					strsize = tmpstr.length();
					tmpstr = tmpstr.substring(2,strsize-1); 
					locctr += tmpstr.length()/2; //char개수만큼 locctr 증가
				}
			}
		}
		return locctr;
	}
	
	/**
	 * index에 해당하는 literal을 리턴한다.
	 * @param index : 리턴할 literal의 인덱스
	 * @return : literal
	 * 
	 */
	public String getLiteral(int index) {
		return literalList.get(index);
	}
	
	/**
	 * index를 이용하여 literal을 찾은 후 해당하는 literald의 address을 리턴한다.
	 * @param index : 리턴할 address의 인덱스
	 * @return : address
	 * 
	 */
	public int getaddress(int index) {
		return locationList.get(index);
	}
	
	/**
	 * literaltable의 크기을 리턴한다.
	 * @return : literalable의 크기
	 * 
	 */
	public int getSize() {
		return literalList.size();
	}
	
}
